source/Halibut/Transport/Proxy/HttpProxyClient.cs

```diff
@@ -347,25 +347,30 @@ async Task SendConnectionCommandAsync(string host, int port, CancellationToken c
             // send the connect request
             await stream.WriteAsync(request, 0, request.Length, cancellationToken);
 
-            // wait for the proxy server to respond
-            await WaitForDataAsync(stream, cancellationToken);
-
             // PROXY SERVER RESPONSE
             // =======================================================================
             //HTTP/1.0 200 Connection Established<CR><LF>
             //[.... other HTTP header lines ending with <CR><LF>..
             //ignore all of them]
             //<CR><LF>    // Last Empty Line
 
-            // create an byte response array  
-            var response = new byte[TcpClient.ReceiveBufferSize];
             var sbuilder = new StringBuilder();
 
-            do
+            var existingTimeout = stream.ReadTimeout;
+            stream.ReadTimeout = WAIT_FOR_DATA_TIMEOUT;
+            try
             {
-                var bytes = await stream.ReadAsync(response, 0, TcpClient.ReceiveBufferSize, cancellationToken);
-                sbuilder.Append(Encoding.UTF8.GetString(response, 0, bytes));
-            } while (stream.DataAvailable);
+                var response = new byte[1]; // Read 1 byte at a time to prevent over read. (This is not efficient but this is only done once per stream)
+                while (!sbuilder.ToString().EndsWith("\r\n\r\n"))
+                {
+                    var bytes = await stream.ReadAsync(response, 0, response.Length, cancellationToken);
+                    sbuilder.Append(Encoding.UTF8.GetString(response, 0, bytes));
+                }
+            }
+            finally
+            {
+                stream.ReadTimeout = existingTimeout;
```

Comment:
A decorator to safely apply this timeout change and then revert would be nice ;)

Reply:
I thought you would never ask.